diff -u qmail-1.03/qmail-smtpd.c qmail-1.03-checkenv/qmail-smtpd.c
From https://marc.info/?l=qmail&m=105641953707239&w=2
--- qmail-1.03/qmail-smtpd.c	1998-06-15 06:53:16.000000000 -0400
+++ qmail-1.03-checkenv/qmail-smtpd.c	2003-06-21 15:58:21.000000000 -0400
@@ -19,6 +19,8 @@
 #include "env.h"
 #include "now.h"
 #include "exit.h"
+#include "fork.h"
+#include "wait.h"
 #include "rcpthosts.h"
 #include "timeoutread.h"
 #include "timeoutwrite.h"
@@ -45,10 +47,15 @@
 void die_read() { _exit(1); }
 void die_alarm() { out("451 timeout (#4.4.2)\r\n"); flush(); _exit(1); }
 void die_nomem() { out("421 out of memory (#4.3.0)\r\n"); flush(); _exit(1); }
+void die_tempfail() { out("421 temporary envelope failure (#4.3.0)\r\n"); flush(); _exit(1); }
+void die_fork() { out("421 Unable to fork (#4.3.0)\r\n"); flush(); _exit(1); }
+void die_exec() { out("421 Unable to exec (#4.3.0)\r\n"); flush(); _exit(1); }
+void die_childcrashed() { out("421 Aack, child crached. (#4.3.0)\r\n"); flush(); _exit(1); }
 void die_control() { out("421 unable to read controls (#4.3.0)\r\n"); flush(); _exit(1); }
 void die_ipme() { out("421 unable to figure out my IP addresses (#4.3.0)\r\n"); flush(); _exit(1); }
 void straynewline() { out("451 See http://pobox.com/~djb/docs/smtplf.html.\r\n"); flush(); _exit(1); }
 
+void err_permfail() { out("553 permanent envelope failure (#5.7.1)\r\n"); flush(); _exit(1); }
 void err_bmf() { out("553 sorry, your envelope sender is in my badmailfrom list (#5.7.1)\r\n"); }
 void err_nogateway() { out("553 sorry, that domain isn't in my list of allowed rcpthosts (#5.7.1)\r\n"); }
 void err_unimpl() { out("502 unimplemented (#5.5.1)\r\n"); }
@@ -222,6 +229,29 @@
 stralloc mailfrom = {0};
 stralloc rcptto = {0};
 
+void checkenv()
+{
+  int child;
+  int wstat;
+  char *checkenvarg[] = { "bin/qmail-checkenv", mailfrom.s, addr.s, 0 };
+
+  switch(child = fork())
+  {
+    case -1:
+      die_fork();
+    case 0:
+      execv(*checkenvarg,checkenvarg);
+      die_exec();
+  }
+
+  wait_pid(&wstat,child);
+  if (wait_crashed(wstat))
+    die_childcrashed();
+  if (wait_exitcode(wstat) == 0) return;
+  if (wait_exitcode(wstat) == 100) err_permfail();
+  die_tempfail();
+}
+
 void smtp_helo(arg) char *arg;
 {
   smtp_greet("250 "); out("\r\n");
@@ -261,6 +291,7 @@
   if (!stralloc_cats(&rcptto,"T")) die_nomem();
   if (!stralloc_cats(&rcptto,addr.s)) die_nomem();
   if (!stralloc_0(&rcptto)) die_nomem();
+  if (env_get("CHECKENV")) checkenv();
   out("250 ok\r\n");
 }
